### DISCUSSION TITLE
USB2RS485 question + 2-speed pump question

### DISCUSSION THREAD

**Discussion by Sazerak** [aqualinkd/AqualinkD#238] [Q&A] [answered]
Hi all, I recently got a pool installed with a Jandy heater. I'm hoping to get AqualinkD setup but I wanted to confirm a couple of things before proceeding. I read through a few sections of the Wiki and searched through the discussions a bit, but I still have some questions/concerns. Sorry in advance for the novel. I've recently come to learn that I'm terrible at being concise (yay for ADHD)...

I'll start by mentioning that my setup is pretty basic compared to (I assume) most people. It's a small 18' above ground pool with a 2-speed 1.5hp Hayward pump (with no automation/timer on it whatsoever) and the Jandy heater is a JXI200NK (standalone, no external control panel).

From what I'm reading here, I think it should be rather simple to hook-up my heater to a Pi with AqualinkD, but I want to confirm a few details to make sure I'm not overlooking anything.

First and foremost, my heater only works when my pump is set to HIGH. Right now, the only way to control that is with the switch at the back of the pump, but since I plan on setting up AqualinkD, I figured I could automate the heater too using a couple of relays (one for on-off and another for low-high) that will be controlled via GPIO. I noticed there's an icon on AqualinkD for a Pump, but I'm wondering if that can be coded to use GPIO or if it works via RS485 (admittedly I haven't read through the code yet). Also, can I add a button to control the Low-High function of my pump?

My next questions are about the USB2RS485 adapter. I plan on getting the recommended one, but I read in the wiki and in a discussion about the Ground and the Terminator wires, but I'm still confused as to the actual recommendation on these. My understanding is that technically I would only need to plug the 2 data wires into the heater, but that I should connect the ground too (I mean, that makes sense right?). Am I understanding this right? It just seems odd to me to even consider not connecting the ground wire when it's right there, so maybe I'm missing something.. As for the Terminator wires, I'm completely confused on those, so if something can offer an ELI5, I'd greatly appreciate it.

This next question should be an easy one: since the recommended USB2RS485 adapter has a USB A but the RPi Zero (which is recommended) only has a Micro-B, how does everyone connect theirs? The obvious answer would be a Micro-B to A adapter, but that would make everything bulky and lopsided, so I'm just curious to hear if some people are doing things differently. This is really just a question out of curiosity though since I plan on using an RPi 3A+ which has a USB A on it.

My last question is about the status of the heater when using AqualinkD. It's my understanding that when using a "remote" (which I assume would include AqualinkD), the control panel on the heater itself can't be used. So how does that work when it comes to errors? For example, if I forget to turn my pump to High (or if I'm overdue for a backwash) and try turning on the heater, I get an error for Low Flow and can't do anything until I turn the heater off and on again (not actually cut power to it, but rather turn the heating function off and back on). So would I see these error codes on AqualinkD, and would I be able to "reset" them by cycling the heater from AqualinkD?

Thanks in advance for any guidance!

**Comment by sfeakes** [comment, marked as answer]
From your description, it doesn’t sound like AqualinkD will work for your setup.  AqualinkD communicates directly with a Jandy control panel. While AqualinkD can read messages from all RS485 devices it does not control them directly. Each RS485 bus needs a master device, (their can only be one master) and that controls communication with every other device on the bus. This master is the control Jandy panel.  So unless you have a Jandy control panel that you have not mentioned above, AqualinkD will not work.

Their are other projects out their that can act as a master device for Pumps and SWG, but none that I know of for JX(i) heaters.   At some point in the future I’ve always imagined creating another project that will replace the control panel entirely, but that’s a long way of at present.

Some of your other questions.
In non RS485 mode, a Jandy control panel will control the heater in exactly the manor you descride. Set heater to max, control panel monitors water temp and turns heater on/off via relay.  In RS485 mode the control panel will simply tell the heater to turn on and set temp to X.  This is usually preferred for a hot-tub situation so you get a constant heat rather than burst of hot water every time the heater cycles.

AqualinkD can not control GPIO directly.  But that would be a fairly simply addition if there was a need.

On the recommended USB2RS485 adapter, some people use an adapter for the USB A to B, some simply use a different style USB2RS485 adapter, or a “full size” pi.

RS485 Bus Termination can be confusing as pool equipment has been hit n miss as to what devices are and are not terminated, so without looking at every RS485 device AND where the USB2RS485 adapter is located it’s hard to say, which is why the wiki tries to explain what needs to happen on termination but can’t give guidance.   But I do knowthe JX(i) heaters are terminated, so if it’s just that and an RS4852USB adapter on the bus, then the adapter would also need to be terminated.

As the the errors you describe, that’s why the control panel is an important part since they have all the safe guards / error checking built in that stop you damaging equipment. If you are controlling devices without the control panel, that’s where you would need to make sure your code / whatever device has all the appropriate safe guards / checking going built in.

**Reply by Sazerak** [reply]
Well darn! I completely missed the fact that AqualinkD was a slave, not a master. I kept seeing references to the Jandy control panels, but I figured it was because most people already had those. I didn't realize it was a requirement.

Back to the drawing board I guess. If I have time, maybe I'll try to poke around with the protocol to see what I can find out (on top of the information already available here and there on the internet).

I'm also looking into the possibility of using the "dumb" connector, but apparently I'm pretty dumb too since I can't figure out how it works. As far as I can tell, I should be able to at least turn the heater on and off with the system in TSTAT mode, but trying to find information on what I'm supposed to wire in there (is it just a switch, or am I providing voltage, etc..) is darn near impossible.

Thanks again for the detailed reply, I really appreciate that you took the time to explain all that! With any luck, I might be able to piece together some things in the future to make myself a basic setup that I can control remotely.

**Reply by sfeakes** [reply]
If using the TSTAT connection, I seem to remember that's just a make / break contact switch.  ie no voltage, just connect the two wires through the contact side of a relay (or switch).

I did have the protocol documented, but had to remove that information.  So you probably won't find anything complete on the net, but there are parts of information on the net.  Pentair and Jandy protocols are different. But the Pentair (and other) SWG use the Jandy protocol. 
This URL described the Pentair protocol, but for Pentair SWG they are actually describing how the Jandy protocol works.  So this is a good start, then you can obviously sift through the code in this repo.  https://github.com/tagyoureit/nodejs-poolController/wiki/Chlorinator

The other thing you can try is search the net for "jandy RS simulator download", this will give you a control panel simulator (ie RS485 master device) that you can run on a windows PC. If you use the RS4852USB connector with the PC and appropriate COM port driver that will control your JXi heater over the RS485 bus.  I wouldn't recommend running this as a permanent solution as the software really is a bad and full of bugs, but it'll at least get you started in understanding how you can control the JXi heater.

**Reply by sfeakes** [reply]
BTW, If you get the PC to JX(i) heater working, and you are serious about writing something yourself. You could then use the serial_logger in this repo with a Raspberry PI (and another RS4852USB adapter) to show you how the PC simulator is controlling the heater.  the serial_logger in this repo is designed to show you everything that's happening on the RS485 bus.

**Reply by Sazerak** [reply]
Thanks for the advice! I did actually find a copy of the protocol so at least I have that to get started with. I also started reading through a "Wiki" that someone wrote where I originally found out about AqualinkD and they seem to have documented some of the basic functions in their "Notes" section. It looks like they've pretty well only documented on/off and check temp, but that's pretty well all I'm looking for, so I might give that a try on my end when I get things setup and go from there.

Here's a link to the page: https://wiki.jmehan.com/display/KNOW/Jandy+Pool+Heater

I really do like these kind of projects and I'm exciting at learning more about the Jandy Protocol, but realistically it might be a while before I can really dig deep into it with young kids at home, but if I do end up making some progress, I'll be sure to keep you in the loop!
